packages/ai/core/tool/mcp/types.ts

```diff
@@ -0,0 +1,251 @@
+   *
+   * This can be used by clients to improve the LLM's understanding of available tools, resources, etc. It can be thought of like a "hint" to the model. For example, this information MAY be added to the system prompt.
+   */
+  instructions: z.optional(z.string()),
+});
+
+const CursorSchema = z.string();
+const PaginatedResultSchema = ResultSchema.extend({
+  nextCursor: z.optional(CursorSchema),
+});
+export const PaginatedRequestSchema = RequestSchema.extend({
+  params: BaseRequestParamsSchema.extend({
+    cursor: z.optional(CursorSchema),
+  }).optional(),
+});
+
+const ToolSchema = z
+  .object({
+    name: z.string(),
+    description: z.optional(z.string()),
+    inputSchema: z
+      .object({
+        type: z.literal('object'),
+        properties: z.optional(z.object({}).passthrough()),
+      })
+      .passthrough(),
+  })
+  .passthrough();
+export const ListToolsResultSchema = PaginatedResultSchema.extend({
+  tools: z.array(ToolSchema),
+});
+export type ListToolsResult = z.infer<typeof ListToolsResultSchema>;
+const ListToolsRequestSchema = PaginatedRequestSchema.extend({
+  method: z.literal('tools/list'),
+});
+export type ListToolsRequest = z.infer<typeof ListToolsRequestSchema>;
+
+const CallToolRequestSchema = RequestSchema.extend({
+  method: z.literal('tools/call'),
+  params: BaseRequestParamsSchema.extend({
+    name: z.string(),
+    arguments: z.optional(z.record(z.unknown())),
+  }),
+});
+export type CallToolRequest = z.infer<typeof CallToolRequestSchema>;
+const TextContentSchema = z
+  .object({
+    type: z.literal('text'),
+    text: z.string(),
+  })
+  .passthrough();
+const ImageContentSchema = z
+  .object({
+    type: z.literal('image'),
+    data: z.string().base64(),
+    mimeType: z.string(),
+  })
+  .passthrough();
+const ResourceContentsSchema = z
+  .object({
+    /**
+     * The URI of this resource.
+     */
+    uri: z.string(),
+    /**
+     * The MIME type of this resource, if known.
+     */
+    mimeType: z.optional(z.string()),
+  })
+  .passthrough();
+const TextResourceContentsSchema = ResourceContentsSchema.extend({
+  text: z.string(),
+});
+const BlobResourceContentsSchema = ResourceContentsSchema.extend({
+  blob: z.string().base64(),
+});
+const EmbeddedResourceSchema = z
+  .object({
+    type: z.literal('resource'),
+    resource: z.union([TextResourceContentsSchema, BlobResourceContentsSchema]),
+  })
+  .passthrough();
+export const CallToolResultSchema = ResultSchema.extend({
+  content: z.array(
+    z.union([TextContentSchema, ImageContentSchema, EmbeddedResourceSchema]),
+  ),
+  isError: z.boolean().default(false).optional(),
+});
+export const CompatibilityCallToolResultSchema = CallToolResultSchema.or(
+  ResultSchema.extend({
+    toolResult: z.unknown(),
+  }),
+);
+export type CallToolResult = z.infer<typeof CallToolResultSchema>;
+export type CompatibilityCallToolResult = z.infer<
+  typeof CompatibilityCallToolResultSchema
+>;
```

Comment:
is all of this needed or can parts be removed? do we need zod schemas for everything?

Reply:
re-reviewing types to see what's _really_ needed!